File: problems/141._inked_List_Cycle/141.Linked_List_Cycle.cpp
#include <unordered_set>
struct ListNode
{
	int val;
	ListNode *next;
	ListNode(int x) : val(x), next(NULL) {}
	ListNode(int x, ListNode *node) : val(x), next(node) {}
};

// @lc code=start
class Solution
{
public:
	bool hasCycle(ListNode *head)
	{
		if (head == NULL || head->next == NULL)
		{
			return false;
		}

		ListNode *rabbit = head->next;
		ListNode *turtle = head;

		while (rabbit != turtle)
		{
			if (rabbit->next == NULL || rabbit->next->next == NULL)
			{
				return false;
			}

			rabbit = rabbit->next->next;
			turtle = turtle->next;
		}
		return true;
	}
};
// @lc code=end

File: problems/141._inked_List_Cycle/141.Linked_List_Cycle.java
class Solution {
	public boolean hasCycle(ListNode head) {
		if (head == null) {
			return false;
		}

		var slow = head;
		var fast = head.next;
		while (slow != null && fast != null) {
			if (slow == fast) {
				return true;
			}

			slow = slow.next;

			fast = fast.next;
			if (fast == null) {
				return false;
			}
			fast = fast.next;
		}

		return false;
	}
}
